src/js.c

```diff
@@ -2827,6 +2903,206 @@ js_PullSubscribe(natsSubscription **sub, jsCtx *js, const char *subject, const c
+    int     remainingUnrequested    = 0;
+    int64_t remainingBytes          = 0;
+    int     want                    = 0;
+    bool    maybeMore               = true;
+
+    nats_lockSubAndDispatcher(sub);
+
+    int isAhead = fetch->requestedMsgs - fetch->deliveredMsgs;
+    int wantAhead = fetch->opts.KeepAhead;
+    if (isAhead > wantAhead)
+        maybeMore = false;
+
+    if (maybeMore)
+    {
+        // fetch->maxMessages is always > 0
+        remainingUnrequested = fetch->opts.MaxMessages - fetch->requestedMsgs;
+        if (remainingUnrequested <= 0)
+            maybeMore = false;
+    }
+
+    if (maybeMore && (fetch->opts.MaxBytes > 0))
+    {
+        remainingBytes = fetch->opts.MaxBytes - fetch->receivedBytes;
+        if (remainingBytes <= 0)
+            maybeMore = false;
+    }
+
+    if (maybeMore)
+    {
+        want = remainingUnrequested;
+        if (want > fetch->opts.FetchSize)
+            want = fetch->opts.FetchSize;
+
+        maybeMore = (want > 0);
+    }
+
+    nats_unlockSubAndDispatcher(sub);
+
+    if (!maybeMore)
+        return false;
+
+    // Since we do not allow keepAhead with MaxBytes, this is an accurate count
+    // of how many more bytes we expect.
+    *maxBytes = remainingBytes;
+    *messages = want;
+    return true;
+}
+
+natsStatus
+js_PullSubscribeAsync(natsSubscription **newsub, jsCtx *js, const char *subject, const char *durable,
+                      natsMsgHandler msgCB, void *msgCBClosure,
+                      jsOptions *jsOpts, jsSubOptions *opts, jsErrCode *errCode)
+{
+    natsStatus          s       = NATS_OK;
+    natsSubscription    *sub    = NULL;
+    jsSub               *jsi    = NULL;
+    jsFetch             *fetch  = NULL;
+
+    if ((newsub == NULL) || (msgCB == NULL))
+        return nats_setDefaultError(NATS_INVALID_ARG);
+
+    if ((jsOpts != NULL) && (jsOpts->PullSubscribeAsync.MaxBytes > 0) && (jsOpts->PullSubscribeAsync.KeepAhead > 0))
+        return nats_setError(NATS_INVALID_ARG, "%s", "Can not use MaxBytes and KeepAhead together");
+
+    if (errCode != NULL)
+        *errCode = 0;
+
+    // Do a basic pull subscribe first, but with a callback so it is treated as
+    // "async" and assigned to a dispatcher. Since we don't fetch anything, it
+    // will not be active yet.
+    s = _subscribe(&sub, js, subject, durable, msgCB, msgCBClosure, true, jsOpts, opts, errCode);
+    if(s == NATS_OK)
+    {
+        fetch = NATS_CALLOC(1, sizeof(jsFetch));
+        if (fetch == NULL)
+            s = nats_setDefaultError(NATS_NO_MEMORY);
+    }
+    if (s != NATS_OK)
+    {
+        NATS_FREE(fetch);
+        natsSubscription_Destroy(sub);
+        return NATS_UPDATE_ERR_STACK(s);
+    }
+
+    // Initialize fetch parameters.
+    fetch->status = NATS_OK;
+    fetch->startTimeMillis = nats_Now();
+
+    if (jsOpts != NULL)
+        fetch->opts = jsOpts->PullSubscribeAsync;
+    if (fetch->opts.FetchSize == 0)
```

Comment:
In the previous code (where you used the _set() macro), was your intent to say that for non "0/NULL" values of the options (if options itself was not NULL), you would use them as-is, if not set some default values? Correct?

I believe this is what you are doing here, but wanted to make sure.

Reply:
Clarified and removed the confusing defaulting here. Still need to ensure the default for FetchSize and Nextf